Random assignment of availability-chunk indices to validators (#47)
polkadot-fellows/RFCs: text/0047-random-assignment-of-availability-chunks.md

1. it's critical to the consensus protocol that all validators have a common view of the Validator->Chunk mapping.
1. it enables further atomic changes to the shuffling algorithm.
1. it enables alternative client implementations (in other languages) to use it
1. mitigates the problem of third-party libraries changing the implementations of the `ChaCha8Rng` or the `rand::shuffle`
that could be introduced in further versions, which would stall parachains. This would be quite an "easy" attack.

Additionally, so that client code is able to efficiently get the mapping from the runtime, another API will be added
for retrieving chunk indices in bulk for all validators at a given block and core.

#### Upgrade path

Considering that the Validator->Chunk mapping is critical to para consensus, the change needs to be enacted atomically via
governance, only after all validators have upgraded the node to a version that is aware of this mapping.
It needs to be explicitly stated that after the runtime upgrade and governance enactment, validators that run older client versions that don't
support this mapping will not be able to participate in parachain consensus.

> ### Getting access to core_index in different subsystems

Reviewer: I think this section is all implementation details of the Parity Polkadot node and should not be here.

Reviewer: I mean, just rephrasing to not refer to a particular architecture (subsystems).

Author: I rewrote the content in this section to refer to the protocol instead of the subsytems of the parity polkadot node

in order to query the runtime API and retrieve the core index that the candidate was occupying.

The usage of the systematic data availability recovery feature will also be subject to all nodes using the V2 disputes networking protocol.

#### Alternatives to using core_index

As an alternative to core_index, the `ParaId` could be used. It has the advantage of being readily available in the
`CandidateReceipt`, which would enable the dispute communication protocol to not change and would simplify the implementation.
However, in the context of [CoreJam](https://github.com/polkadot-fellows/RFCs/pull/31), `ParaId`s will no longer exist (at least not in their current form).

## Drawbacks

Has a fair amount of technical complexity involved:

- Introduces another runtime API that is going to be issued by multiple subsystems. With adequate client-side caching, this should be acceptable.

- Requires a networking protocol upgrade on the disputes request-response protocol

## Testing, Security, and Privacy

Extensive testing will be conducted - both automated and manual.
This proposal doesn't affect security or privacy.

## Performance, Ergonomics, and Compatibility

### Performance

> This is a necessary optimisation, as reed-solomon erasure coding has proven to be a top consumer of CPU time in polkadot when scaling the parachain count.

Reviewer: ```suggestion
This is a necessary DA optimisation, as reed-solomon chunk reconstruction has proven to be a top consumer of CPU time as we scale up the parachain block size and number of availability cores.
```

> With this optimisation, preliminary performance results show that CPU time used for reed-solomon coding can be halved and total POV recovery time decrease by 80% for large POVs. See more [here](https://github.com/paritytech/polkadot-sdk/issues/598#issuecomment-1792007099).

Reviewer: Calling into the runtime to get chunk mappings is slower than doing that in the client, especially doing so many calls and building up overhead. I suggest only a bulk API, so it's just one call per candidate.

Author: There are places (like bitfield signing) that don't need to query the chunk indices of all the validators, since we only need to attest that we have our chunk.

That's why I think it's valuable to have both a bulk API and one for retrieving a single index

per session, naively using the ValidatorIndex as the ChunkIndex would pose an unreasonable stress on the first N/3
validators during an entire session, when favouring availability recovery from systematic chunks.

Therefore, the relay chain node needs a deterministic way of evenly distributing the first ~(N_VALIDATORS / 3)
systematic availability chunks to different validators, based on the session, relay chain block number and core.
The main purpose is to ensure fair distribution of network bandwidth usage for availability recovery in general and in
particular for systematic chunk holders. 

## Stakeholders

Relay chain node core developers.

## Explanation

### Systematic erasure codes

An erasure coding algorithm is considered systematic if it preserves the original unencoded data as part of the
resulting code.
[The implementation of the erasure coding algorithm used for polkadot's availability data](https://github.com/paritytech/reed-solomon-novelpoly) is systematic. Roughly speaking, the first N_VALIDATORS/3
chunks of data can be cheaply concatenated to retrieve the original data, without running the resource-intensive and

> time-consuming reconstruction algorithm.

Author: I added the concatenation algorithm to the doc

order. If this fails, fallback to option (b).

(b) for large PoVs (over 128 Kib), launch N parallel requests for the erasure coded chunks (currently, N has an upper
limit of 50), until enough chunks were recovered. Validators are tried in a random order. Then, reconstruct the
original data.

### Availability recovery from systematic chunks

As part of the effort of
[increasing polkadot's resource efficiency, scalability and performance](https://github.com/paritytech/roadmap/issues/26),
work is under way to modify the Availability Recovery subsystem by leveraging systematic chunks. See
[this comment](https://github.com/paritytech/polkadot-sdk/issues/598#issuecomment-1792007099) for preliminary
performance results.

In this scheme, the relay chain node will first attempt to retrieve the N/3 systematic chunks from the validators that
should hold them, before falling back to recovering from regular chunks, as before.

### Chunk assignment function

#### Properties
The function that decides the chunk index for a validator should be parametrised by at least
`(validator_index, session_index, block_number, core_index)`

> and have the following properties:

Reviewer 2: in other words what want is: create a uniformly distributed but deterministic mapping (shuffle) ValidatorIndex -> ChunkIndex per block per core

Author: yes. I'll add this

1. deterministic
1. pseudo-random
1. relatively quick to compute and resource-efficient.
1. when considering the other params besides `validator_index` as fixed,
the function should describe a random permutation of the chunk indices

> 1. considering `session_index` and `block_number` as fixed arguments, the validators that map to the first N/3 chunk indices should

Reviewer 2: Thinking more about my argument to use `session_index` [here](https://github.com/paritytech/polkadot-sdk/issues/598#issuecomment-1734075024), I think it's a very niche edge-case that's not worth worrying about. An alternative to (`session_index`, `block_number`) would be block hash which would also sidestep that issue.

Author: Sounds fair. Another issue I just thought of with using block number is that for disputes hapenning on unimported forks, the ChainAPI call for getting the block number would also fail.

have as little overlap as possible for different cores.

> #### Proposed function and runtime API

Reviewer 2: I think we should either propose a runtime API or the (client-side) function implementation and I'm leaning towards the latter

Author: maybe it wasn't clear. I was only proposing the runtime API here

Pseudocode:

```rust
pub fn get_chunk_index(
    n_validators: u32,
    validator_index: ValidatorIndex,
    session_index: SessionIndex,
    block_number: BlockNumber,
    core_index: CoreIndex
) -> ChunkIndex {
    let threshold = systematic_threshold(n_validators); // Roughly n_validators/3
    let seed = derive_seed(session_index, block_number);
    let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);
    let mut chunk_indices: Vec<ChunkIndex> = (0..n_validators).map(Into::into).collect();
    chunk_indices.shuffle(&mut rng);

    let core_start_pos = threshold * core_index.0;
    return chunk_indices[(core_start_pos + validator_index) % n_validators]
}
```

> The function should be implemented as a runtime API, because:

Reviewer 2: tbh, only reason 2 seems convincing to me and client features runtime API makes it less convincing

> 1. it's critical to the consensus protocol that all validators have a common view of the Validator->Chunk mapping.

Reviewer 2: so is approval-voting, which is off-chain

and in this case, it's critical for finality + parachain liveness, not safety

Author: fair enough. this bullet-point should be coerced with the one talking about the supply chain attack

Author: this discussion is obsolete, as the current proposal does not include a runtime API for the mapping function

1. it enables further atomic changes to the shuffling algorithm.
1. it enables alternative client implementations (in other languages) to use it
1. mitigates the problem of third-party libraries changing the implementations of the `ChaCha8Rng` or the `rand::shuffle`

> that could be introduced in further versions, which would stall parachains. This would be quite an "easy" attack.

Reviewer 2: I didn't get this argument. Are you talking about supply chain attacks? How is it specific to this RFC?

Author: here's an example:

1. all validators are running normally, using systematic recovery.
2. a contributor on the `rand` crate (malicious or not) makes a perfectly valid change to the shuffling algorithm, that results in a different output for `shuffle()`.
3. there's a polkadot node release that bumps rand to this new version.
4. some of the validators upgrade their node.
5. as a result, some/all parachains are stalled, because some validators have a differing view of the validator->chunk mapping. Also, the validators that upgraded stop receiving rewards

IMO this has a high chance of happening in the future.

I view this mapping in a similar way to the per-session validator shuffle we do in the runtime to choose the validator active set.

Author: now, we could implement our own version of `shuffle`, which would mitigate this issue.
The thing I'm concerned about is implementing our own version of `ChaCha8Rng`. How feasible is it to assume that it won't change in the future? CC: @burdges

Reviewer 3: It's not a supply chain attack so much as rand having a history of instability.

1. The `pov_recovery` task of the collators starts availability recovery in response to noticing a candidate getting
backed, which enables easy access to the core index the candidate started occupying.
1. Disputes may be initiated on a number of occasions:
  
    3.a. is initiated by the validator as a result of finding an invalid candidate while participating in the
  approval-voting protocol. In this case, availability-recovery is not needed, since the validator already issued their
  vote.
    
    3.b is initiated by the validator noticing dispute votes recorded on-chain. In this case, we can safely
  assume that the backing event for that candidate has been recorded and kept in memory.
  
    3.c is initiated as a result of getting a dispute statement from another validator. It is possible that the dispute
  is happening on a fork that was not yet imported by this validator, so the subsystem may not have seen this candidate
  being backed.

As a solution to 3.c, a new version for the disputes request-response networking protocol will be added.
This message type will include the relay block hash where the candidate was included. This information will be used

> in order to query the runtime API and retrieve the core index that the candidate was occupying.

Author: on a second thought: is this even possible? If the validator is not aware of the fork, how can it call a runtime API on that fork?

Reviewer 2: if the validator hasn't seen that fork, it's can't call into runtime API of that

The usage of the systematic data availability recovery feature will also be subject to all nodes using the V2 disputes
networking protocol.

#### Alternatives to using core_index

As an alternative to core_index, the `ParaId` could be used. It has the advantage of being readily available in the
`CandidateReceipt`, which would enable the dispute communication protocol to not change and would simplify the
implementation.
However, in the context of [CoreJam](https://github.com/polkadot-fellows/RFCs/pull/31), `ParaId`s will no longer exist

> (at least not in their current form).

Reviewer 2: IMHO we shouldn't worry too much about CoreJam compatibility. IIUC there might be a drop-in replacement for `ParaId` (`AuthId`), so we should first explore an avenue with `ParaId` perhaps. 

The (bigger) problem with `ParaId` is that it's claimable by users, so in the theory one could create a collision attack where multiple paras have the same systematic chunks. In practice, I believe such an attack would be high cost and low benefit. And, perhaps, it could be mitigated by using a cryptographic hash of `ParaId` (or `AuthId` or `core_index`) in the future.

Reviewer 3: We should use the core index here probably, not the para id.  I'd thought the core indexes could be spaced somewhat evenly, but with the actual sequence being random.

Reviewer 3: I changed my view here above in https://github.com/polkadot-fellows/RFCs/pull/47#discussion_r1433047970  

It's probably more future proof if we do not require the core index here, so the map uses num_cores, num_validators, relay parent, and paraid.  We've other options but this looks pretty future proof.  CoreJam cannot really work without something like paraid.

Author: Using paraid was my first attempt, but it complicates things quite a bit.
As Andronik notes, it's claimable by a user. I don't think it's the biggest problem though, they can't get a different ParaId every day (can they?).

`paraid`'s biggest problem is that it doesn't form a strictly monontonically increasing sequence, so we'd have to do something else with it. My initial attempt was to seed a ChaCha8 RNG with it and generate one random number. Then this number was the offset into the chunk index vector.

But then if we complicate it so much, we may want to embed it into the runtime, so that future updates to the `rand` crate don't break us (which creates all sorts of problems that could be avoided).

We've had this paraid -> core_index back and forth for a while now, it'd be great if we could all (the ones interested in this RFC) hop on a call or somehow reach a conclusion. There are pros and cons to both, AFAICT.

I think this remains the only bit that prevents this RFC from moving on (anybody correct me if I'm wrong)

The main purpose is to ensure fair distribution of network bandwidth usage for availability recovery in general and in
particular for systematic chunk holders. 

## Stakeholders

Relay chain node core developers.

## Explanation

### Systematic erasure codes

An erasure coding algorithm is considered systematic if it preserves the original unencoded data as part of the
resulting code.
[The implementation of the erasure coding algorithm used for polkadot's availability data](https://github.com/paritytech/reed-solomon-novelpoly) is systematic.
Roughly speaking, the first N_VALIDATORS/3 chunks of data can be cheaply concatenated to retrieve the original data,
without running the resource-intensive and time-consuming reconstruction algorithm.

Here's the concatenation procedure of systematic chunks for polkadot's erasure coding algorithm
(minus error handling, for briefness):
```rust
pub fn reconstruct_from_systematic<T: Decode>(
	n_validators: usize,
	chunks: Vec<&[u8]>,
) -> T {
	let mut threshold = (n_validators - 1) / 3;
	if !is_power_of_two(threshold) {
		threshold = next_lower_power_of_2(threshold);
	}

	let shard_len = chunks.iter().next().unwrap().len();

> 	let mut systematic_bytes = Vec::with_capacity(shard_len * kpow2);

Reviewer 2: ```suggestion
	let mut systematic_bytes = Vec::with_capacity(shard_len * threshold);
```

	for i in (0..shard_len).step_by(2) {

> 		for chunk in chunks.iter().take(kpow2) {

Reviewer 2: ```suggestion
		for chunk in chunks.iter().take(threshold) {
```

			systematic_bytes.push(chunk[i]);
			systematic_bytes.push(chunk[i + 1]);
		}
	}

	Decode::decode(&mut &systematic_bytes[..]).map_err(|err| Error::Decode(err))
}

> ```

Reviewer 3: This could be done in the reed-solomn crate itself.

Author: done: See https://github.com/paritytech/reed-solomon-novelpoly/pull/36

> In a nutshell, it performs a column-wise concatenation with 2-bit chunks.

Reviewer 2: ```suggestion
In a nutshell, it performs a column-wise concatenation with 2-byte chunks.
```

order. If this fails, fallback to option (b).

(b) for large PoVs (over 128 Kib), launch N parallel requests for the erasure coded chunks (currently, N has an upper
limit of 50), until enough chunks were recovered. Validators are tried in a random order. Then, reconstruct the
original data.

### Availability recovery from systematic chunks

As part of the effort of
[increasing polkadot's resource efficiency, scalability and performance](https://github.com/paritytech/roadmap/issues/26),
work is under way to modify the Availability Recovery protocol by leveraging systematic chunks. See
[this comment](https://github.com/paritytech/polkadot-sdk/issues/598#issuecomment-1792007099) for preliminary
performance results.

In this scheme, the relay chain node will first attempt to retrieve the ~N/3 systematic chunks from the validators that
should hold them, before falling back to recovering from regular chunks, as before.

### Chunk assignment function

#### Properties

The function that decides the chunk index for a validator should be parameterized by at least
`(validator_index, relay_parent, para_id)`
and have the following properties:
1. deterministic

> 1. pseudo-random

Author: agreed, I removed the randomness part

1. relatively quick to compute and resource-efficient.
1. when considering the other params besides `validator_index` as fixed, the function should describe a random permutation
of the chunk indices
1. considering `relay_parent` as a fixed argument, the validators that map to the first N/3 chunk indices should
have as little overlap as possible for different paras scheduled on that relay parent.

In other words, we want a uniformly distributed, deterministic mapping from `ValidatorIndex` to `ChunkIndex` per block
per scheduled para.

#### Proposed runtime API

The mapping function should be implemented as a runtime API, because:

1. it enables further atomic changes to the shuffling algorithm.
1. it enables alternative client implementations (in other languages) to use it
1. considering how critical it is for parachain consensus that all validators have a common view of the Validator->Chunk
mapping, this mitigates the problem of third-party libraries changing the implementations of the `ChaCha8Rng` or the `rand::shuffle`

> that could be introduced in further versions. This would stall parachains if only a portion of validators upgraded the node.

Reviewer 2: We should obviously not rely in consensus on something that is not specified to be deterministic, like https://polkadot.network/blog/a-polkadot-postmortem-24-05-2021#the-good. And try to reduce third-party dependencies in general.
I would assume that a library that implements `ChaCha8Rng` adheres to the spec, otherwise it's a bug.
To mitigate supply-chain issues including bugs, we should probably use `cargo-vet` or a similar tool, but again, this is out of scope of and not limited to this RFC.

Reviewer 4: This is off-topic for this RFC, but as a heads up we already do this: https://github.com/paritytech/polkadot-sdk/blob/2d09e83d0703ca6bf6aba773e80ea14576887ac7/polkadot/node/network/gossip-support/src/lib.rs#L601-L610

Pseudocode:

```rust
pub fn get_chunk_index(
  n_validators: u32,
  validator_index: ValidatorIndex,
  relay_parent: Hash,
  para_id: ParaId
) -> ChunkIndex {
  let threshold = systematic_threshold(n_validators); // Roughly n_validators/3
  let seed = derive_seed(relay_parent);

>   let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);

Reviewer 3: Avoid paraid here imho, not overly keen on relay parent either.

Instead use era/session, slot, and chain spec to define the validator sequence, and then core index to define the start position in the validator sequence.

>   let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);

Reviewer 5: Validators are already randomized at the beginning of the session. Core index for start position makes sense. What is wrong with using the block number in addition? 

Reason, I would like to have it dependent on the block (could also be slot, I just don't see the benefit) is that by having a start position by core index, we ensure equal distribution of systemic chunks across a block, but paras are not all equal. Some could be heavier than others, hence it would be beneficial to change the mapping each block. 

In my opinion we could also use the hash instead of the block number, here I think anything is likely better than static.

>   let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);

Reviewer 3: We want slot or block I think.  We're not overly adversarial here, but you can manipulate block numbers easily, which slot numbers represent an opportunity for doing something, so you always pay something to chose another one.  I'd say slot not block.

We randomize the validator list based upon the randomness two epochs/sessions ago?  Cool.  Any idea if we similarly randomize the map from paraids to cores per era/session too?  If yes, then maybe we could just progress sequentially through them?  
```
let k = num_validators / num_cores;
let fist_validator_index_for_core = ((core_id - slot) * k % num_validators) as u32;
```
We'd prefer the randomization of core_id for this because otherwise you could still make hot spots.  We could've hot spots even with this scheme, but not so easy to make them.  We'd avoid those if we randomize per slot.

Also this exactly computation does not work due to signed vs unsigned, and the fact that it suggests things progress backwards as time progresses time, which again tried to avoid hot spots.

>   let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);

Author: > Any idea if we similarly randomize the map from paraids to cores per era/session too?

AFAICT from the scheduler code, we don't (at least for the parachain auction model; for on-demand, I see more complicated logic which takes into account core affinities).

> We're not overly adversarial here, but you can manipulate block numbers easily, while slot numbers represent an opportunity for doing something, so you always pay 19 DOTs or more to chose another one. I'd say slot not block.

I'm having a hard time understanding the advantage of slot number vs block number. This may be simply because I don't know that much about slots. AFAICT, slots are equally useful as block number for the mapping function (monotonically increasing by 1), except that there may be slots that go unoccupied (if the chain is stalled for example) and are therefore skipped. Is that correct?

> so you always pay 19 DOTs or more to chose another one

what is this fee? where can I read more about this?

>   let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);

Reviewer 5: @burdges I don't get your arguments. How can block numbers be manipulated? They are always increasing by 1, you clearly must be talking about forks. So an adversary could create multiple forks, with all the same load distribution and tries to overload validators this way? If we create forks or reversions we already have a performance problem anyway. 

Really not getting how slot numbers are better here. I also don't get your argument about hot spots, my proposal above was precisely to avoid hot spots (by not using randomness). What do you mean by hot spots and how would randomness help here?

A single validator not providing its systemic chunk would be enough to break systemic recovery. I don't see how randomization schemes help here. If we were smart we could somehow track the validators withholding systemic chunks and then make an assignment where they are all pooled together into one candidate. This way, at least only one systemic recovery fails. (We could equally well just remove them entirely.)

But honestly, I would not worry too much about this here. If we ever found that validators try to mess with this on purpose, the threat is low enough that social (calling them out) and governance would be adequate measures to deal with them.

>   let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);

Author: > A single validator not providing its systemic chunk would be enough to break systemic recovery

Slightly offtopic: In my WIP PR I added functionality to request up to 5 chunks from the backing group as a backup solution, so that a couple of validators not returning their systematic chunks would not invalidate the entire procedure

>   let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);

Reviewer 3: We could permit requesting them all from the backing group, but make the availability providers the first choice, meaning maybe:  1st, try all systemic chunk providers.  2nd, try remaining systemic chunks from backers.  3rd, fetch random non-systemic chunks.  The concern is just that we overload the backers.

It'll also make rewards somewhat more fragile, but likely worth the difficulty for the performance.

for retrieving chunk indices in bulk for all validators at a given block and core:

```rust
pub fn get_chunk_indices(
  n_validators: u32,
  relay_parent: Hash,
  para_id: ParaId
) -> Vec<ChunkIndex> {
  let threshold = systematic_threshold(n_validators); // Roughly n_validators/3
  let seed = derive_seed(relay_parent);
  let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);
  let mut chunk_indices: Vec<ChunkIndex> = (0..n_validators).map(Into::into).collect();
  chunk_indices.shuffle(&mut rng);

  let seed = derive_seed(hash(para_id));
  let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);

  let para_start_pos = rng.gen_range(0..n_validators);

  chunk_indices
    .into_iter()
    .cycle()
    .skip(para_start_pos)
    .take(n_validators)
    .collect()
}
```

#### Upgrade path

Considering that the Validator->Chunk mapping is critical to para consensus, the change needs to be enacted atomically
via governance, only after all validators have upgraded the node to a version that is aware of this mapping.
It needs to be explicitly stated that after the runtime upgrade and governance enactment, validators that run older

> client versions that don't support this mapping will not be able to participate in parachain consensus.

Reviewer 5: Does this need to be the case? As discussed, I think they should be able to still do the recovery from non systemic chunks.

Reviewer 3: yeah but they need to know what chunks map to where.

Reviewer 5: yep in distribution, for fetching I think the mapping ValidatorIndex -> ChunkIndex can be unknown, without any real issues.

Additionally, an error will be logged when starting a validator with an older version, after the runtime was upgraded and the feature enabled.

## Drawbacks

- In terms of guaranteeing even load distribution, a simpler function that chooses the per-core start position in the
shuffle as `threshold * core_index` would likely perform better, but considering that the core_index is not part of the
CandidateReceipt, the implementation would be too complicated. More details in [Appendix A](#appendix-a).
- Considering future protocol changes that aim to generalise the work polkadot is doing (like CoreJam), `ParaId`s may be
removed from the protocol, in favour of more generic primitives. In that case, `ParaId`s in the availability recovery
process should be replaced with a similar identifier. It's important to note that the implementation is greatly simplified
if this identifier is part of the `CandidateReceipt` or the future analogous data structure.

> - It's a breaking change that requires most validators to be upgrade their node version.

Reviewer 5: superfluous `be`.

## Prior Art and References

See comments on the [tracking issue](https://github.com/paritytech/polkadot-sdk/issues/598) and the
[in-progress PR](https://github.com/paritytech/polkadot-sdk/pull/1644)

## Unresolved Questions

- Is it the best option to embed the mapping function in the runtime?
- Is there a better upgrade path that would preserve backwards compatibility?
- Is usage of `ParaId` the best choice for spreading out the network load during systematic chunk recovery within the
same block?

## Future Directions and Related Material

This enables future optimisations for the performance of availability recovery, such as retrieving batched systematic
chunks from backers/approval-checkers.

## Appendix A

This appendix explores alternatives to using the `ParaId` as the factor by which availability chunk indices are
distributed to validators within the same relay chain block, and why they weren't chosen.

### Core index

Here, `core_index` refers to the index of the core that a candidate was occupying while it was pending availability
(from backing to inclusion).

Availability-recovery can currently be triggered by the following phases in the polkadot protocol:

> 1. During the approval voting process.

Reviewer 5: As long as recovery is possible in disputes, it should be fine if it can not always use systemic recovery. What is missing for me, is to understand why this needs to be an incompatible change. Shouldn't recovery be possible always, even if you don't know what the systemic chunks are?

Reviewer 2: Currently, when you request a chunk, you need to know which chunk to request. If you specify a wrong chunk, you'll get an empty response. So everyone should be onboard how chunks are shuffled in order for reconstruction to work, not just from systematic chunks. Unless we rely on backers, which we can't in case of a dispute.

We could add an additional request type to request any chunk you have. That could probably be a reasonable fallback in this case.

Author: > We could add an additional request type to request any chunk you have. That could probably be a reasonable fallback in this case.

I don't think this achieves the purpose of availability-recovery working with an arbitrary mix of upgraded & non-upgraded validators. Adding a new request type (or even changing the meaning of the existing one) would also mean a node upgrade (because responding to the new type of request is only useful for the nodes that haven't yet upgraded to use the new mapping). So validators might as well just upgrade to the version that uses the new shuffle. I'm not sure this has any benefit

Reviewer 2: I meant as a fallback in case we want to use `core_id`s which might be not readily available in some cases of disputes.

Author: > I meant as a fallback in case we want to use core_ids which might be not readily available in some cases of disputes.

I see. In this case, it would be a reasonable fallback. But this would mean giving up systematic recovery in those cases. I'm wondering if it's worth doing all of this just to replace the para-id thingy with the core_id

Reviewer 5: As discussed in chat. If we keep requesting via `ValidatorIndex`, then in recovery you don't need to know the mapping, if you are fine with non systemic recovery.

Reviewer 3: You need to know the chunk-validator mapping.  we may say the chunk index in the chunk itself, but you'd then think you could still get the same chunk for the guy who just gave you one.  In extreme cases, I bet this breaks reconstruction.

Reviewer 5: Indeed, we discussed this. The only issue is that you could not know whether you got the chunk you actually asked for. But that is only relevant if we wanted to change reputations. With that aside, we can just check whether we already fetched that chunk - if yes, treat it as invalid/garbage and move on to the next validator.

Reviewer 3: It's worse than reputations.  I think it changes whether the 2/3rds honest assumption works correctly.

Author: > It's worse than reputations. I think it changes whether the 2/3rds honest assumption works correctly.

Hm, how?

we currently don't modify any reputations within the availability-recovery subsystem. also, what @eskimor is describing above can happen with the current production code. the node does not check that it received the chunk it requested. It only checks that it's a valid chunk and that the proof is valid

Reviewer 3: The first 1/3rd of validators refuse to talk to you.  You obtain some chunks from most of the second 1/3rd, but they give you the chunks of the third 1/3rd, so you keep asking the second 1/3rd for the last chunks, which they do not have.  Reconstruction fails.

Author: > You obtain some chunks from most of the second 1/3rd, but they give you the chunks of the third 1/3rd

Doesn't that mean the second third is also malicious? They don't adhere to the network protocol (which in polkadot's case is still specified by the reference implementation AFAICT).

We assume that (aside from the malicious third), all validators will respond with unique chunks. We don't care that they're not the ones they should/shouldn't hold, as long as they don't overlap. I think this is a reasonable assumption, isn't it? (we are already making this assumption in the current released code).

Leaving systematic recovery aside, the key aspect would be that we wouldn't ask for the chunk by its index. we would ask each validator for their chunk. The responders would map their validator index to the right chunk index. They would then respond with that chunk. Assuming at most a third of them are malicious and they all have the same mapping, we should get at least two thirds correct responses (and we only need a third. And for non-systematic recovery, we don't care if the responses are actually the ones we asked for, as long as we don't get duplicates.

For systematic recovery, we'd assume the validator initiating recovery has the mapping and requests chunks from the validators. Systematic recovery will then check that the chunks received are indeed the ones requested and fall back to regular behaviour if a peer misbehaves.

Author: > No. We're discussing if it's okay for validators to have different mappings. I'm saying no.

I was discussing the case where validators all have the same mapping, but a validator participating in a dispute does not have access to the mapping at that point.

Reviewer 3: Ahh fine.  We'll loose the mappings sometime before the data expires?  We should probably avoid doing that.  We can obviously ask everyone and account for the DLs differently, but I'd prefer to be more like bittorrent trackers here, so that parachains can fetch data this way.

Reviewer 5: Yes. All validators will know the correct mapping, if they have seen (and still know) the relay parent. There is a good chance that this is the case even in disputes, but it is not guaranteed and recovery must work even in that case (but can be slower).

Using the core index should be fine, what makes that even more acceptable to me is that we actually want (and need to) include the core index in the candidate receipt at some point. So the moment we have this change, then the block availability will cease to matter.
